flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java

```diff
@@ -89,17 +101,98 @@ public void checkpointState(
 			checkpointMetaData.getCheckpointId(),
 			checkpointOptions.getTargetLocation());
 
-		CheckpointingOperation.execute(
-			checkpointMetaData,
-			checkpointOptions,
-			checkpointMetrics,
-			storage,
-			operatorChain,
-			taskName,
-			closeableRegistry,
-			executorService,
-			env,
-			asyncExceptionHandler,
-			isCanceled);
+		long startSyncPartNano = System.nanoTime();
+
+		HashMap<OperatorID, OperatorSnapshotFutures> operatorSnapshotsInProgress = new HashMap<>(operatorChain.getNumberOfOperators());
+		try {
+			for (StreamOperatorWrapper<?, ?> operatorWrapper : operatorChain.getAllOperators(true)) {
+				StreamOperator<?> op = operatorWrapper.getStreamOperator();
+				OperatorSnapshotFutures snapshotInProgress = checkpointStreamOperator(
+					op,
+					checkpointMetaData,
+					checkpointOptions,
+					storage,
+					isCanceled);
+				operatorSnapshotsInProgress.put(op.getOperatorID(), snapshotInProgress);
+			}
+
+			if (LOG.isDebugEnabled()) {
+				LOG.debug("Finished synchronous checkpoints for checkpoint {} on task {}",
+					checkpointMetaData.getCheckpointId(), taskName);
+			}
+
+			long startAsyncPartNano = System.nanoTime();
+
+			checkpointMetrics.setSyncDurationMillis((startAsyncPartNano - startSyncPartNano) / 1_000_000);
+
+			// we are transferring ownership over snapshotInProgressList for cleanup to the thread, active on submit
+			executorService.execute(new AsyncCheckpointRunnable(
+				operatorSnapshotsInProgress,
+				checkpointMetaData,
+				checkpointMetrics,
+				startAsyncPartNano,
+				taskName,
+				closeableRegistry,
+				env,
+				asyncExceptionHandler));
+
+			if (LOG.isDebugEnabled()) {
+				LOG.debug(
+					"{} - finished synchronous part of checkpoint {}. Alignment duration: {} ms, snapshot duration {} ms",
+					taskName, checkpointMetaData.getCheckpointId(),
+					checkpointMetrics.getAlignmentDurationNanos() / 1_000_000,
+					checkpointMetrics.getSyncDurationMillis());
+			}
+		} catch (Exception ex) {
+			// Cleanup to release resources
+			for (OperatorSnapshotFutures operatorSnapshotResult : operatorSnapshotsInProgress.values()) {
+				if (null != operatorSnapshotResult) {
+					try {
+						operatorSnapshotResult.cancel();
+					} catch (Exception e) {
+						LOG.warn("Could not properly cancel an operator snapshot result.", e);
+					}
+				}
+			}
+
+			if (LOG.isDebugEnabled()) {
+				LOG.debug(
```

Comment:
I see this log occur twice with the same arguments only different message, and this log actually seems a bit long to impact the normal logics review. I am not sure whether it is worth extracting a separate method for only passing different message in two usages.

Reply:
I refactored a bit `SubtaskCheckpointCoordinatorImpl` while addressing one of the issues below and these messages ended up one after another; so removed one.